doc/api/quic.md

```diff
@@ -0,0 +1,2276 @@
+# QUIC
+
+<!--introduced_in=REPLACEME-->
+
+> Stability: 1 - Experimental
+
+The `net` module provides an implementation of the QUIC protocol. To
+access it:
+
+```js
+const { createQuicSocket } = require('net');
+```
+
+## Example
+
+```js
+'use strict';
+
+const key = getTLSKeySomehow();
+const cert = getTLSCertSomehow();
+
+const { createQuicSocket } = require('net');
+
+// Create the QUIC UDP IPv4 socket bound to local IP port 1234
+const socket = createQuicSocket({ endpoint: { port: 1234 } });
+
+// Tell the socket to operate as a server using the given
+// key and certificate to secure new connections, using
+// the fictional 'hello' application protocol.
+socket.listen({ key, cert, alpn: 'hello' });
+
+socket.on('session', (session) => {
+  // A new server side session has been created!
+
+  session.on('secure', () => {
+    // Once the TLS handshake is completed, we can
+    // open streams...
+    const uni = session.openStream({ halfOpen: true });
+    uni.write('hi ');
+    uni.end('from the server!');
+  });
+
+  // The peer opened a new stream!
+  session.on('stream', (stream) => {
+    // Let's say hello
+    stream.end('Hello World');
+
+    // Let's see what the peer has to say...
+    stream.setEncoding('utf8');
+    stream.on('data', console.log);
+    stream.on('end', () => console.log('stream ended'));
+  });
+});
+
+socket.on('listening', () => {
+  // The socket is listening for sessions!
+});
+```
+
+## QUIC Basics
+
+QUIC is a UDP-based network transport protocol that includes built-in security
+via TLS 1.3, flow control, error correction, connection migration,
+multiplexing, and more.
+
+Within the Node.js implementation of the QUIC protocol, there are three main
+components: the `QuicSocket`, the `QuicSession` and the `QuicStream`.
+
+### QuicSocket
+
+A `QuicSocket` encapsulates a binding to one or more local UDP ports. It is
+used to send data to, and receive data from, remote endpoints. Once created,
+a `QuicSocket` is associated with a local network address and IP port and can
+act as both a QUIC client and server simultaneously. User code at the
+JavaScript level interacts with the `QuicSocket` object to:
+
+* Query or modified the properties of the local UDP binding;
+* Create client `QuicSession` instances;
+* Wait for server `QuicSession` instances; or
+* Query activity statistics
+
+Unlike the `net.Socket` and `tls.TLSSocket`, a `QuicSocket` instance cannot be
```

Comment:
Using `QuicSocket` because it is conceptually consistent with `net.Socket`, `tls.Socket`, etc already in use in core. There is also already a distinction between a `QuicSocket` and `QuicEndpoint` in that a single `QuicSocket` can actually be bound to multiple local `QuicEndpoint` instances simultaneously -- that is, a `QuicSocket` can listen on and transmit via multiple local bound UDP ports.

Reply:
I understand the relationship, but I think that net.Socket is more similar to QuicEndpoint. But I'm OK with current names.